### ISSUE TITLE
PBRGraph shader size in memory is too much - 433.2 MBs

### ISSUE THREAD

**Issue by umeraamir-arthur** [KhronosGroup/UnityGLTF#792]
Hi!

I was wondering why this PBRGraph shader has this much size in memory even though if I've no GLBs in the scene. I can surely place any GLB at runtime but currently at the time of this memory capture there was no GLB present in the scene.

![image](https://github.com/user-attachments/assets/3a35cdd8-6df4-43f5-bf60-1bf3f47e8f99)

I've included both UnityGLTF shaders in my project Graphics settings because we don't have any GLB in the scene but we can place any type of GLB at runtime.

![image](https://github.com/user-attachments/assets/a43cb354-d559-45b4-b9a4-38bed0590c79)

We don't support draco, GPU instancing or these things so I was wondering if there is any way that I could remove some parts from this shader referring to these things and my apps memory will be saved in some way?

Any suggestion would be appreciated, Thankyou
cc. @pfcDorn @hybridherbst

**Comment by hybridherbst** [comment]
Hey @umeraamir-arthur, instead of adding the shaders directly to the "Always Included" list, use the Shader Variant Collections that ship with UnityGLTF. 

By adding them to "Always Included" you are including ALL potential variants, for any combination of Unity rendering features, including ones that Unity would otherwise strip. 
By using the Shader Variant Collections, you still get full runtime support, but only the variants that you need (e.g. the same ones that Unity also creates for their own shaders).

**Comment by umeraamir-arthur** [comment]
Yes and by adding directly to "Always Included" shaders we usually have to wait for 3 hours for first build to create them in Library folder and people hates me for that :D
Can you point me to the location where Shader Variant Collections are placed in UnityGLTF. I'm not able to see them. Maybe giving an example will be amazing e.g. this is the shader variant which is used when you've GPU instancing in your GLBs. This is the shader variant when you've KTX2 in your GLBs, and this one is for Draco...... Maybe a little like that...

**Comment by hybridherbst** [comment]
They’re here: https://github.com/KhronosGroup/UnityGLTF/tree/main/Runtime/Shaders/VariantCollections

There included in the latest release already. There’s no extra variant sets for different features, these should „just work“ and absolutely not be as big as what you saw with „Always Include“.

**Comment by umeraamir-arthur** [comment]
That's Great!

We're using URP with forward render pass. Is this the right settings that I selected?
![image](https://github.com/user-attachments/assets/96eba904-e1ba-4f38-9444-76a02e900b77)

**Comment by robLE83** [comment]
Hey, yes...that's correct 👍

**Comment by umeraamir-arthur** [comment]
Thanks @robLE83 !!!
And a last question before creating a build.... the steps are
1) Remove PBRGraph and UnlitGraph from Always Include Shaders
2) Just create a build
I think I don't need to change anything for Shader Variant Collections to initialize. Rest of the things are handled by UnityGLTF... right? For both shaders i.e. PBRGraph and UnlitGraph?

**Comment by umeraamir-arthur** [comment]
Any suggestions on how to remove this shader completely if there is no GLB in the scene. And as soon as we place any GLB in the room then it can load the shader, maybe?

**Comment by pfcDorn** [comment]
Yeah, that's "by design". This is to avoid loading/compiling a shader when you instantiate a model which is using such a shader. Otherwise you would first see a pink material until the shader is ready for rendering (or you would have frame drops in this moment). Not perfect for low-end platforms, but i'm not aware of any solutions :/

**Comment by umeraamir-arthur** [comment]
Yes that would be an issue. I remember once I was profiling GLTFast and there shader was adaptive from scene to scene. On one scene it was ~100MB and on the other scene it was ~200MB.... Idk what were they using never explored it after then.

**Comment by hybridherbst** [comment]
I believe actually Unity 6 supports dynamic shader loading – haven't tested
it yet though.
